compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -3553,26 +3563,139 @@ private STNode parseVarDeclRhs(STNode metadata, STNode finalKeyword, STNode type
                     expr = parseActionOrExpression();
                 }
                 semicolon = parseSemicolon();
+                hasVarInit = true;
                 break;
             case SEMICOLON_TOKEN:
                 assign = STNodeFactory.createEmptyNode();
                 expr = STNodeFactory.createEmptyNode();
                 semicolon = parseSemicolon();
                 break;
             default:
-                recover(nextToken, ParserRuleContext.VAR_DECL_STMT_RHS, metadata, finalKeyword, typedBindingPattern,
-                        isModuleVar);
-                return parseVarDeclRhs(metadata, finalKeyword, typedBindingPattern, isModuleVar);
+                recover(nextToken, ParserRuleContext.VAR_DECL_STMT_RHS, metadata, isolatedFinalQualifiers,
+                        typedBindingPattern, isModuleVar);
+                return parseVarDeclRhs(metadata, isolatedFinalQualifiers, typedBindingPattern, isModuleVar);
         }
 
         endContext();
         if (isModuleVar) {
-            return STNodeFactory.createModuleVariableDeclarationNode(metadata, finalKeyword, typedBindingPattern,
-                    assign, expr, semicolon);
+            if (hasVarInit || isolatedFinalQualifiers.isEmpty()) {
+                STNode isolatedFinalQualifierNode = STNodeFactory.createNodeList(isolatedFinalQualifiers);
+                return STNodeFactory.createModuleVariableDeclarationNode(metadata, isolatedFinalQualifierNode,
+                        typedBindingPattern, assign, expr, semicolon);
+            }
+
+            // Only final qualifier is allowed for non initialized module var declaration
+
+            // If type descriptor in the binding pattern is either object or function type
+            // and qualifier list has isolated qualifier as the last token,
+            // that isolated qualifier should be part of the type.
+            STNode lastQualifier = isolatedFinalQualifiers.get(isolatedFinalQualifiers.size() - 1);
+            if (lastQualifier.kind == SyntaxKind.ISOLATED_KEYWORD) {
+                lastQualifier = isolatedFinalQualifiers.remove(isolatedFinalQualifiers.size() - 1);
+                typedBindingPattern =
+                        modifyTypedBindingPatternWithMissingQualifier(typedBindingPattern, lastQualifier);
```

Comment:
Method name `modifyTypedBindingPatternWithMissingQualifier` suggests you are updating the node with a missing qualifier, but here the qualifier is `isolated`